### PR TITLE
`<algorithm>`: Implement worst-case linear-time `nth_element`

### LINKED ISSUE #856: <algorithm>: nth_element does not comply with worst case running time requirements
**Describe the bug**
The standard requires `nth_element` (both `std` and `std::ranges` overloads) to perform only O(n lg n) swaps. However, we currently implement only quickselect: https://github.com/microsoft/STL/blob/ff947563da6511270d8b7292e162f8549df8267f/stl/inc/algorithm#L4470

This has worst case running time O(n^2) if we always choose bad pivots, just like quicksort. We need some form of 'ideal' checking similar to how `std::sort` works, and fall back to a [median-of-medians](https://en.wikipedia.org/wiki/Median_of_medians) or similar implementation.

**Additional context**

This was previously recorded in a Microsoft-internal bug database but I closed the issue there as I had filed it against myself.

### PR DESCRIPTION
**Pull request by muellerj2**
Fixes #856 for `std::nth_element` and `std::ranges::nth_element`. This implements a fallback to the median-of-medians-of-five algorithm when the quickselect algorithm seems to be making too little progress.

The median-of-medians algorithm is mostly the textbook version, with two minor tweaks:
* If the processed sequence doesn't cleanly divide into groups of five elements, the remainder group with less than five elements isn't considered for the median computation. (This reduces the amount of code and doesn't make any difference in the asymptotics. I couldn't observe any practical difference in running time, too.)
* When the pivot (=median-of-medians) has been computed, all (greater) medians located after the pivot are moved to the very end of the processed sequence and the pivot is swapped into the middle of the sequence. This is because all of these elements are guaranteed to be moved by the pivot partitioning algorithm, so this step immediately moves them into an appropriate position (or the pivot probably closer to it). This way, the medians can also be excluded from the sequence on which the partitioning algorithm is applied, avoiding some unnecessary comparisons. (In practice, the benchmarks suggested that this makes the algorithm a few percent faster, but the difference is minor.)

### Benchmark results

`bm_uniform` just applies `nth_element` to an integer array of the given length. The integer array is uniformly sampled from a fixed seed. This is to check that the worst-case fallback does not noticeably worsen the processing time on such a sequence.

`bm_tunkey_adversary` applies `nth_element` to a sequence on which the implemented quickselect algorithm performs terribly.

Before:
```
--------------------------------------------------------------------------------
Benchmark                                      Time             CPU   Iterations
--------------------------------------------------------------------------------
bm_uniform<alg_type::std_fn>/1024           1845 ns         1803 ns       407273
bm_uniform<alg_type::std_fn>/2048           3966 ns         3990 ns       172308
bm_uniform<alg_type::std_fn>/4096           7702 ns         7673 ns        89600
bm_uniform<alg_type::std_fn>/8192          18090 ns        18032 ns        40727
bm_uniform<alg_type::rng>/1024              1759 ns         1758 ns       373333
bm_uniform<alg_type::rng>/2048              3985 ns         4011 ns       179200
bm_uniform<alg_type::rng>/4096              7694 ns         7847 ns        89600
bm_uniform<alg_type::rng>/8192             18015 ns        17997 ns        37333
bm_tunkey_adversary<alg_type::std_fn>      12995 ns        13393 ns        56000
bm_tunkey_adversary<alg_type::rng>         12714 ns        12835 ns        56000
```

After:
```
--------------------------------------------------------------------------------
Benchmark                                      Time             CPU   Iterations
--------------------------------------------------------------------------------
bm_uniform<alg_type::std_fn>/1024           1599 ns         1604 ns       448000
bm_uniform<alg_type::std_fn>/2048           3626 ns         3610 ns       194783
bm_uniform<alg_type::std_fn>/4096           7068 ns         7150 ns        89600
bm_uniform<alg_type::std_fn>/8192          16469 ns        16044 ns        44800
bm_uniform<alg_type::rng>/1024              1701 ns         1709 ns       448000
bm_uniform<alg_type::rng>/2048              3841 ns         3931 ns       194783
bm_uniform<alg_type::rng>/4096              7447 ns         7324 ns        74667
bm_uniform<alg_type::rng>/8192             17024 ns        16741 ns        37333
bm_tunkey_adversary<alg_type::std_fn>       6075 ns         5929 ns        89600
bm_tunkey_adversary<alg_type::rng>          6270 ns         6278 ns       112000
```

As expected, the fallback greatly improves the running time for `bm_tunkey_adversary`. The timings for `bm_uniform` are about on par, or more precisely even a bit better with this PR on my machine.

### The fallback heuristic

`std::sort` switches to its fallback when the recursion depth exceeds some logarithmic threshold. We could use the same heuristic as well, however, this would not guarantee linear time in the worst case but "only" an $O(n \log n)$ bound. Alternatively, we could limit the recursion depth to some constant, but that's likely a pessimization for large sequences.

So I opted for an adaptive depth limit: Like the heuristic for `std::sort`, it assumes that each iteration should reduce the range of inspected elements by 25 %.  But while `std::sort` derives a maximum recursion depth from this assumption, this heuristic falls back to the median-of-medians algorithm when the actual size of the processed sequence exceeds the desired size by some constant tolerance factor (currently about 2) during some iteration. Thus, the total number of processed elements over all quickselect iterations is bounded by a multiple of the sequence length times a geometric sum, ensuring worst-case linear time overall. At the same time, the tolerance factor introduces some leeway so that one or two bad iterations (especially at the beginning) don't trigger the fallback immediately.

Obviously, there are many possible choices for the desired percentage reduction per iteration and the tolerance factor. But the benchmarks seem to suggest that the chosen values aren't too bad; a smaller percentage reduction or a larger margin factor noticeably worsen the `bm_tunkey_adversary` benchmark, but result in little difference for `bm_uniform`.  Besides, the implementation of `std::sort` already sets a precedent for a desired reduction of 25 % per iteration.

### Test

The newly added test applies `nth_element` to the same worst-case sequence as the benchmark. This makes sure that the fallback is actually exerted by the test. (I think it's also the first test that exerts the quickselect algorithm and not the just the insertion sort fallback.)

### CODE REVIEW

**Comment by muellerj2** [comment]
> There's a systematic typo - the algorithm we use is called "Tukey's ninther", not Tunkey.

Oops. Fixed.

File: stl/inc/algorithm [hunk 1]
```diff
@@ -9320,6 +9333,114 @@ namespace ranges {
...
+            return;
+        }
+
+        _Diff _Median_count = _Length / static_cast<_Diff>(5);
+        auto _Median_last   = _First + _Median_count;
+        auto _Current       = _First;
+        for (auto _Median_dest = _First; _Median_dest != _Median_last; ++_Median_dest) {
+            _STD _Swap_med5_unchecked(_Current, _Median_dest, _Pred);
+            _Current += static_cast<_Diff>(5);
+        }
```
Comment (davidmrdavid) [inline stl/inc/algorithm:9386]: This bit took me a bit to understand because I did not expect the medians to be collected by putting them in positions `_Current`, then `_Current + 1`, `_Current + 2`, and so on until `_Current + _Median_count`. I assumed the medians would be placed in their "expected position" if the array were sorted; that's on me :-) .

 _A subjective nit_ based on the above - I'd appreciate a short comment at the beginning of this block indicating that we'll accumulate the medians "at the beginning of the container". Perhaps @StephanTLavavej can help me improve the phrasing here, in case it can be made more accurate.

File: stl/inc/algorithm [hunk 2]
```diff
@@ -9320,6 +9333,114 @@ namespace ranges {
...
+
+        // the medians before and after _Median_mid are
+        // less equal and greater equal _Median_mid, respectively.
+        // move the ones after _Median_mid near _Last and then partition
+        // the elements between the medians at front and back by the pivot _Median_mid
+        _Diff _Median_greater_count = (_Median_last - _Median_mid) - 1;
+        auto _Median_greater_first  = _Last - _Median_greater_count;
+        _STD _Swap_ranges_unchecked(_Median_mid + static_cast<_Diff>(1), _Median_last, _Median_greater_first);
+        auto _Middle = _First + ((_Last - _First) >> 1);
+        swap(*_Median_mid, *_Middle);
```
Comment (davidmrdavid) [inline stl/inc/algorithm:9399]: nit - I think we should add an empty newline between the swapping of ranges and the declaration of `_Middle`. The swapping of ranges ends an optimization step (moving the greater elements near `_Last` and the declaration of `_Middle` starts a new logical step - the partitioning step.

For that reason, I think it would be good to separate these visually as well.

File: stl/inc/algorithm [hunk 3]
```diff
@@ -9320,6 +9333,114 @@ namespace ranges {
...
+            _Current += static_cast<_Diff>(5);
+        }
+
+        auto _Median_mid = _First + (_Median_count >> 1);
+        _STD _Nth_element_median_of_medians(_First, _Median_mid, _Median_last, _Pred);
+
+        // the medians before and after _Median_mid are
+        // less equal and greater equal _Median_mid, respectively.
+        // move the ones after _Median_mid near _Last and then partition
+        // the elements between the medians at front and back by the pivot _Median_mid
```
Comment (davidmrdavid) [inline stl/inc/algorithm:9394]: nit - lit would be good for the comment to explain why we're moving the elements near `_Last`.

It's a bit surprising (to me, at least), so I think would be good to call out that this is an optimization (and why it is one).

**Review by davidmrdavid — COMMENTED** [review]
Thanks a lot for this PR ⭐.
Left some comments, mostly nits that I think would improve readability, and some questions.

File: stl/inc/algorithm [hunk 4]
```diff
@@ -9320,6 +9333,114 @@ namespace ranges {
...
+        _Diff _Median_count = _Length / static_cast<_Diff>(5);
+        auto _Median_last   = _First + _Median_count;
+        auto _Current       = _First;
+        for (auto _Median_dest = _First; _Median_dest != _Median_last; ++_Median_dest) {
+            _STD _Swap_med5_unchecked(_Current, _Median_dest, _Pred);
+            _Current += static_cast<_Diff>(5);
+        }
+
+        auto _Median_mid = _First + (_Median_count >> 1);
+        _STD _Nth_element_median_of_medians(_First, _Median_mid, _Median_last, _Pred);
```
[the remark this answers is not in the document]
Comment (muellerj2) [reply stl/inc/algorithm:9389]: There is no easy way to avoid the recursion. The textbook median-of-medians select algorithm performs two recursive calls: One to compute the median-of-medians-of-5, and a tail-recursive call on the range that includes the nth element after partitioning the sequence. This implementation translates the tail call into a loop, but the first recursive call can't be easily turned into a loop because the local variables are accessed again when the recursive call returns. We would have to simulate the call stack on the heap if we tried to eliminate this recursion.

However, it is guaranteed that the length of the processed range in the recursive call is a fifth of the length of the original call. So for a sequence of length $m$, we have are most $\log_5(m/32)$ recursive calls (because at 32 elements remaining, the algorithm switches to insertion sort). So for a range of length $2^{64}$, this performs at most $26$ recursive calls. That less than half of the worst case of `std::sort()`, for example, which performs up to $\log_{2}(m/32)$ recursive calls (so about $59$ for an input sequence of length $2^{64}$).

File: tests/std/tests/GH_000856_nth_element_linear/test.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    for (vector<int>::difference_type nth{}, last = src.end() - src.begin(); nth < last;
+         nth += static_cast<vector<int>::difference_type>(15)) {
+        v = src;
+        alg(v, nth);
+        check_result(sorted, v, nth);
+    }
+    v = src;
+    alg(v, (v.end() - v.begin()) - static_cast<vector<int>::difference_type>(1));
+    check_result(sorted, v, (v.end() - v.begin()) - static_cast<vector<int>::difference_type>(1));
```
[the remark this answers is not in the document]
Comment (muellerj2) [reply tests/std/tests/GH_000856_nth_element_linear/test.cpp:106]: It's checking that we get the result right for the ultimate (actual) element in the range. (`v.end() - v.begin()` is the length of the array, `v.end() + ((v.end() - v.begin()) - 1)` is the iterator pointing to the last element in the sequence.)

As for the reason: To make sure we get the result right at the boundary condition. The prior loop checks every 15th element. But the last element is not at a position divisible by 15 for the specific inputs the test currently uses.

I guess we could also add a test that `nth_element` doesn't crash when the requested nth element is `v.end()` (as there isn't much to check about the result itself). But the median-of-medians fallback algorithm is actually not exercised in this case, so I didn't immediately see a need to add this test here. (The test `Dev11_0000000_user_defined_literals` already checks for a larger input that `nth_element` doesn't crash or modify the input sequence).

We could also check every possible position rather than every 15th, if you prefer that. The test will take a bit more time, though.

File: benchmarks/src/nth_element.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    vector<int> src(begin(tukey_ninther_adversary), end(tukey_ninther_adversary));
+
+    vector<int> v;
+    v.reserve(src.size());
+
+    for (auto _ : state) {
+        v = src;
+        benchmark::DoNotOptimize(v);
+        if constexpr (Type == alg_type::std_fn) {
+            nth_element(v.begin(), v.begin() + (v.size() >> 1) + (v.size() >> 2), v.end());
```
[the remark this answers is not in the document]
Comment (muellerj2) [reply benchmarks/src/nth_element.cpp:110]: For the previous benchmark, `v.begin() + (v.size() / 2)` would be one of the more difficult choices for the algorithm in expectation if a truly random sequence was used. (I haven't checked what the actual most difficult choice is for the sequence specifically generated here.)

The same isn't isn't true for the adversary sequence. Without the fallback, the quickselect algorithm roughly performs more work the later the chosen `nth` is in the sequence. That is, the maximum work is performed for `v.end() - 1`, while the minimum processing is done for `v.begin()`. Moreover, `nth` at a later position in the vector lets us better observe how a change to the fallback heuristic affects the algorithm performance (as this means the fallback is triggered for more possible ways we could modify the heuristic).

We could also execute `nth_element` for `v.begin() + (v.size() >> 1)`, though. The main consequence is that it will reduce the speedup we observe in the benchmark.

File: tests/std/tests/GH_000856_nth_element_linear/test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (muellerj2) [reply tests/std/tests/GH_000856_nth_element_linear/test.cpp:106]: But now that I think more about my argument that we should test the boundary condition, this specific adversary sequence doesn't do that for `nth=v.begin()`. I think I should generate a second adversary that has the inverse characteristic: The closer `nth` is to `v.begin()`, the more work is performed. Then we can test the algorithm for both adversaries and thus check that median-of-medians algorithm works correctly at both boundaries.

File: stl/inc/algorithm (hunk 2 above)
Comment (muellerj2) [reply stl/inc/algorithm:9399]: Just a remark to your observation that this is part of the partitioning and not the optimization step: I don't quite agree. From my point of view, it is still mainly part of the optimization step before we do the partitioning.

We could also immediately pass `_Median_mid` to the partition function without moving the median-of-medians to the middle. However, by swapping the median-of-medians, the median-of-medians will probably be moved around much less by the partitioning step.

Specifically, we know that at least 1/5 of the elements are greater than the median-of-medians and 1/5 are less than it, but we don't know where these elements are in the range to be partitioned. So if we leave the median of medians at `_Median_mid`, then it is basically guaranteed that the partition function will move the median-of-medians by at least 1/5 of the sequence through swapping with neighbors, since 1/5 of the elements are smaller than it (exception: there are lots of elements equal to the median-of-medians). If we move the median-of-medians immediately to the middle, then the partition function will move the median-of-medians through swapping with its neighbors by at most 1/5 of the sequence.

File: benchmarks/src/nth_element.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (muellerj2) [reply benchmarks/src/nth_element.cpp:110]: I generated a new adversary for the test case and will add it to the benchmark as well. But it has different characteristics than the previous adversary: quickselect does less work on `v.begin() + (v.size() >> 1) + (v.size() >> 2)` than on `v.begin() + (v.size() >> 1)`.

I don't want to make another special choice for the second adversary, so I will just make all benchmarks find the median of the sequence like `bm_uniform`.

**Comment by muellerj2** [comment]
New benchmarks with new adversary2 and minor changes to benchmark code:

Before:
```
------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations
------------------------------------------------------------------------------------------
bm_uniform<alg_type::std_fn>/1024                     1550 ns         1500 ns       448000
bm_uniform<alg_type::std_fn>/2048                     3413 ns         3376 ns       203636
bm_uniform<alg_type::std_fn>/4096                     6583 ns         6627 ns        89600
bm_uniform<alg_type::std_fn>/8192                    15467 ns        15346 ns        44800
bm_uniform<alg_type::rng>/1024                        1552 ns         1538 ns       497778
bm_uniform<alg_type::rng>/2048                        3494 ns         3299 ns       203636
bm_uniform<alg_type::rng>/4096                        6502 ns         6557 ns       112000
bm_uniform<alg_type::rng>/8192                       15444 ns        15067 ns        49778
bm_tukey_adversary<alg_type::std_fn>/adversary1      11353 ns        11440 ns        56000
bm_tukey_adversary<alg_type::rng>/adversary1         11505 ns        11475 ns        64000
bm_tukey_adversary<alg_type::std_fn>/adversary2      59287 ns        58594 ns        11200
bm_tukey_adversary<alg_type::rng>/adversary2         55013 ns        54688 ns        10000
```

After:
```
------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations
------------------------------------------------------------------------------------------
bm_uniform<alg_type::std_fn>/1024                     1552 ns         1538 ns       497778
bm_uniform<alg_type::std_fn>/2048                     3412 ns         3376 ns       203636
bm_uniform<alg_type::std_fn>/4096                     6622 ns         6417 ns       112000
bm_uniform<alg_type::std_fn>/8192                    15345 ns        15346 ns        44800
bm_uniform<alg_type::rng>/1024                        1458 ns         1444 ns       497778
bm_uniform<alg_type::rng>/2048                        3296 ns         3296 ns       213333
bm_uniform<alg_type::rng>/4096                        6513 ns         6557 ns       112000
bm_uniform<alg_type::rng>/8192                       14922 ns        14753 ns        49778
bm_tukey_adversary<alg_type::std_fn>/adversary1       6043 ns         5999 ns       112000
bm_tukey_adversary<alg_type::rng>/adversary1          5324 ns         5441 ns       112000
bm_tukey_adversary<alg_type::std_fn>/adversary2       4069 ns         4098 ns       179200
bm_tukey_adversary<alg_type::rng>/adversary2          3649 ns         3683 ns       203636
```

**Comment by StephanTLavavej** [comment]
Thanks! I'm getting through my backlog of reviews and will look at this soon.

**Comment by StephanTLavavej** [comment]
I pushed a merge with `main` to pick up toolset and LLVM updates, and to resolve a trivial adjacent-add conflict in `benchmarks/CMakeLists.txt`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! :green_heart: And apologies for the significant delay in getting around to reviewing this!

I pushed minor cleanups to the product code and moderate refactorings to the test/benchmark code. Please meow if I messed anything up.

I reviewed @davidmrdavid's comments and I believe we've addressed them all.

<details><summary>Click to expand 5950X benchmark results:</summary>

Benchmark                                       |   Before |    After | Speedup
------------------------------------------------|----------|----------|--------
`bm_uniform<alg_type::std_fn>/1024`             |  1877 ns |  1801 ns | 1.04
`bm_uniform<alg_type::std_fn>/2048`             |  4190 ns |  3968 ns | 1.06
`bm_uniform<alg_type::std_fn>/4096`             |  8131 ns |  7701 ns | 1.06
`bm_uniform<alg_type::std_fn>/8192`             | 20427 ns | 19078 ns | 1.07
`bm_uniform<alg_type::rng>/1024`                |  1951 ns |  1736 ns | 1.12
`bm_uniform<alg_type::rng>/2048`                |  4197 ns |  3855 ns | 1.09
`bm_uniform<alg_type::rng>/4096`                |  8120 ns |  7487 ns | 1.08
`bm_uniform<alg_type::rng>/8192`                | 25961 ns | 20594 ns | 1.26
`benchmark_common<alg_type::std_fn>/adversary1` | 14572 ns |  6835 ns | 2.13
`benchmark_common<alg_type::rng>/adversary1`    | 15813 ns |  6323 ns | 2.50
`benchmark_common<alg_type::std_fn>/adversary2` | 69962 ns |  4226 ns | 16.56
`benchmark_common<alg_type::rng>/adversary2`    | 69475 ns |  4430 ns | 15.68

</details>

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this major bug that went unfixed for decades by some of the world's most experienced C++ Standard Library implementers! :heart_eyes_cat: :hammer_and_wrench: :green_heart:

**Comment by BillyONeal** [comment]
I wanted to do this for ages, and I'm so happy someone finally has! Thank you so much!
